src/runtime/security/kerberos_utils.cpp

```diff
@@ -0,0 +1,313 @@
+    if (err == KRB5_LNAME_NOTRANS) {
+        if (principal->length > 0) {
+            int cnt = 0;
+            while (cnt < principal->length) {
+                std::string tname;
+                tname.assign((const char *)principal->data[cnt].data, principal->data[cnt].length);
+                if (!username.empty()) {
+                    username += '/';
+                }
+                username += tname;
+                cnt++;
+            }
+            return error_s::make(ERR_OK);
+        }
+        return error_s::make(ERR_KRB5_INTERNAL, "parse username from principal failed");
+    }
+
+    // KRB5_CONFIG_NOTENUFSPACE means BUF_LEN is not enough
+    if (err == KRB5_CONFIG_NOTENUFSPACE) {
+        return error_s::make(ERR_KRB5_INTERNAL, fmt::format("username is larger than {}", BUF_LEN));
+    }
+    KRB5_RETURN_NOT_OK(err, "krb5 parse aname to localname failed");
+
+    if (strlen(buf) <= 0) {
+        return error_s::make(ERR_KRB5_INTERNAL, "empty username");
+    }
+
+    username.assign((const char *)buf);
+    return error_s::make(ERR_OK);
+}
+
+class kinit_context
+{
+public:
+    kinit_context() : _opt(nullptr) {}
+    virtual ~kinit_context();
+    // implementation of 'kinit -k -t <keytab_file> <principal>'
+    error_s kinit(const std::string &keytab_file, const std::string &principal);
+
+private:
+    // krb5 structure
+    krb5_principal _principal;
+    // keytab file with absolute path
+    krb5_keytab _keytab;
+    krb5_ccache _ccache;
+    krb5_get_init_creds_opt *_opt;
+
+    // principal and username that logged in as, this determines "who I am"
+    std::string _principal_name;
+    std::string _username;
+
+    uint64_t _cred_expire_timestamp;
+
+    std::shared_ptr<boost::asio::deadline_timer> _timer;
+
+private:
+    // get _principal_name and _user_name from _principal
+    error_s get_formatted_identities();
+    // get or renew credentials from KDC and store it to _ccache
+    error_s get_credentials();
+    void schedule_renew_credentials();
+};
+
+kinit_context::~kinit_context() { krb5_get_init_creds_opt_free(g_krb5_context, _opt); }
+
+void kinit_context::schedule_renew_credentials()
+{
+    // reserve 300 seconds for renew
+    int64_t renew_gap = _cred_expire_timestamp - utils::get_current_physical_time_s() - 300;
+    if (renew_gap < 300)
+        renew_gap = 300;
+    ddebug_f("schedule to renew credentials in {} seconds later", renew_gap);
+
+    // why don't we use timers in rDSN framework?
+    //  1. currently the rdsn framework may not started yet.
+    //  2. the rdsn framework is used for codes of a service_app,
+    //     not for codes under service_app
+    if (nullptr == _timer) {
+        _timer.reset(new boost::asio::deadline_timer(tools::shared_io_service::instance().ios));
+    }
+    _timer->expires_from_now(boost::posix_time::seconds(renew_gap));
+    _timer->async_wait([this](const boost::system::error_code &err) {
+        if (!err.failed()) {
+            get_credentials();
```

Comment:
Will you try to find credential in krb5_ccache in other PRs?

Reply:
No, the krb5_ccache is used for kerberos internal, we don't need it.

Reply:
It's not internal

Reply:
I have read some docs of kerberos. It says 
```The credential cache file holds Kerberos protocol credentials (for example, tickets, session keys, and other identifying information) in semipermanent storage. The Kerberos protocol reads credentials from the cache as they are required and stores new credentials in the cache as they are obtained. This relieves the application of the responsibility for managing the credentials itself.``` in https://www.ibm.com/support/knowledgecenter/en/SSGSMK_7.1.0/management_sym/sym_kerberos_configuring_credential_cache_file.html

And I have read the corresponding code in kudu, there are no code to try to find credential in krb5_ccache. It just put credentials into krb5_ccache.

Reply:
so you mean krb5_cc_ funcs are useless? You should read more. It's not for internal. I didn't say we need to use it.

Reply:
The Kerberos protocol reads credentials from the cache as they are required and stores new credentials in the cache as they are obtained. I didn't say they are useless

Reply:
So what the credential cache is used for in your opinion? @vagetablechicken

Reply:
What I said is the basic conceptive problem. krb5_ccache is not for krb internal. The owner can use it to avoid unnecessary connections.
There's no need to optimize prematurely.